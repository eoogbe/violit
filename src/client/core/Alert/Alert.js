/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { MdClose } from 'react-icons/md';
import styled from 'styled-components';

const StyledAlert = styled.div`
  font-size: 1rem;
  display: flex;
  align-items: flex-start;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  padding: ${({ theme }) => theme.spacingMd};
  background-color: ${({ theme }) => theme.red200};
`;

const CloseButton = styled.button`
  padding: ${({ theme }) => theme.spacingSm};
  border: none;
  background: transparent;
  color: ${({ theme }) => theme.gray700};
`;

CloseButton.displayName = 'CloseButton';

const Paragraph = styled.p`
  margin: 0;
`;

const ReloadButton = styled.button`
  font-size: 1rem;
  padding: 0;
  border: none;
  background: none;
  text-decoration: underline;
`;

ReloadButton.displayName = 'ReloadButton';

/** An error alert message. */
const Alert = ({ action }) => {
  const [dismissed, setDismissed] = useState(false);

  if (dismissed) {
    return null;
  }

  const closeButtonIconAttrs = {
    title: 'Dismiss alert',
    focusable: 'false',
    viewBox: '0 0 24 24',
  };

  return (
    <StyledAlert role="alert" aria-live="assertive">
      <Paragraph>
        Oops! Something went wrong with {action || 'that action'}!&nbsp;
        <ReloadButton type="button" onClick={() => window.location.reload()}>
          Reload page
        </ReloadButton>
      </Paragraph>
      <CloseButton type="button" onClick={() => setDismissed(true)}>
        <MdClose size="1.25em" attr={closeButtonIconAttrs} />
      </CloseButton>
    </StyledAlert>
  );
};

Alert.propTypes = {
  /** A message describing the attempted action causing the error. */
  action: PropTypes.string,
};

export default Alert;
