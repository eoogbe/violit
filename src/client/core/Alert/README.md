```jsx
<div style={{ position: 'relative', height: '75px' }}>
  <Alert action="creating the foo" />
</div>
```
