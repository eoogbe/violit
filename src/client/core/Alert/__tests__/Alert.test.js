/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { mount } from 'enzyme';
import TestRenderer from 'react-test-renderer';
import Alert from '../Alert';

it('renders without crashing', () => {
  const renderer = TestRenderer.create(<Alert />);
  expect(renderer.toJSON()).toMatchSnapshot();
});

it('renders with an action', () => {
  const renderer = TestRenderer.create(<Alert action="creating the foo" />);
  expect(renderer.toJSON()).toMatchSnapshot();
});

it('when close button clicked is hidden', () => {
  const wrapper = mount(<Alert />);
  wrapper.find('CloseButton').simulate('click');
  expect(wrapper.children()).toHaveLength(0);
});

it('when reload button clicked reloads the page', () => {
  const reload = jest.fn();
  global.location.reload = reload;
  const wrapper = mount(<Alert />);

  wrapper.find('ReloadButton').simulate('click');

  expect(reload).toHaveBeenCalled();
});
