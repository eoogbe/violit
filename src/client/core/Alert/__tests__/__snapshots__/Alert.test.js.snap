// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with an action 1`] = `
<div
  aria-live="assertive"
  className="sc-bdVaJa dXBAhf"
  role="alert"
>
  <p
    className="sc-htpNat bhpHJD"
  >
    Oops! Something went wrong with 
    creating the foo
    ! 
    <button
      className="sc-bxivhb ErYiX"
      onClick={[Function]}
      type="button"
    >
      Reload page
    </button>
  </p>
  <button
    className="sc-bwzfXH haUZkm"
    onClick={[Function]}
    type="button"
  >
    <svg
      fill="currentColor"
      focusable="false"
      height="1.25em"
      size="1.25em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      title="Dismiss alert"
      viewBox="0 0 24 24"
      width="1.25em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
  </button>
</div>
`;

exports[`renders without crashing 1`] = `
<div
  aria-live="assertive"
  className="sc-bdVaJa dXBAhf"
  role="alert"
>
  <p
    className="sc-htpNat bhpHJD"
  >
    Oops! Something went wrong with 
    that action
    ! 
    <button
      className="sc-bxivhb ErYiX"
      onClick={[Function]}
      type="button"
    >
      Reload page
    </button>
  </p>
  <button
    className="sc-bwzfXH haUZkm"
    onClick={[Function]}
    type="button"
  >
    <svg
      fill="currentColor"
      focusable="false"
      height="1.25em"
      size="1.25em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      title="Dismiss alert"
      viewBox="0 0 24 24"
      width="1.25em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
  </button>
</div>
`;
